### PR TITLE
AB#2508 Create signed SBOMs for each build

### PR DESCRIPTION
**Pull request by datosh**
Signed-off-by: Fabian Kammel <fk@edgeless.systems>

### Proposed change(s)

- SBOMs are now:
  - Generated for each build (not just releases).
  - Signed using cosign
  - "attached to container image" and uploaded to ghcr.io. The verification step was added as documentation and test: `cosign verify-attestation ...`
- For CLI's SBOM a detached signature is added to release artifacts.
- Grype is used to automatically fail a build if a vulnerability with risk "high" is found in generated SBOMs.

### Successful manual runs of changed pipelines
Images
* https://github.com/edgelesssys/constellation/actions/runs/3274150228
* https://github.com/edgelesssys/constellation/actions/runs/3274150318
* https://github.com/edgelesssys/constellation/actions/runs/3274150407
* https://github.com/edgelesssys/constellation/actions/runs/3274150491
* https://github.com/edgelesssys/constellation/actions/runs/3274163805

CLI
* https://github.com/edgelesssys/constellation/actions/runs/3274145074

To verify the build verification service from references pipeline run execute (any tag of the same container will also verify successfully. It uses the attached attestation for the corresponding container digest): 
```sh
curl -LO https://github.com/edgelesssys/constellation/releases/download/v2.1.0/cosign.pub
cosign verify-attestation ghcr.io/edgelesssys/constellation/verification-service@sha256:fe31333a0696e4b044a2e1e7c6d6c88d7c387753350f68e7533383e1c70e6360 --type 'https://cyclonedx.org/bom' --key cosign.pub
```

### Checklist

- [x] Update [CHANGELOG.md](https://github.com/edgelesssys/constellation/blob/main/CHANGELOG.md)
- [x] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Link to Milestone

### CODE REVIEW

File: docs/docs/workflows/sbom.md
Comment (daniel-weisse) [inline docs/docs/workflows/sbom.md:80]:
```suggestion
[Grype] is a CLI based tool which lends itself well for integration into CI/CD systems or local developer machines. It's also able to consume the signed attestation statement directly and does the verification in one go.
```

**Review by daniel-weisse — APPROVED** [review]
LGTM, just some minor things annotated by Vale

**Review by malt3 — APPROVED** [review]
LGTM

**Comment by thomasten** [comment]
> Generated for each build (not just releases)

Does this publish something to rekor? Or does any other workflow/action publish something to rekor for non-release builds? We changed it to prevent it once. Not sure if this is still the case.

**Comment by malt3** [comment]
Just as an idea: We could use two separate sets of keys for 

- releases
- testing, staging, development

This way we can test the signatures / sboms for every image. I try to establish a similar method for secure boot signing.

**Comment by datosh** [comment]
> > * Generated for each build (not just releases).
> > * Signed using cosign
> 
> Does this publish something to rekor? Or does any other workflow/action publish something to rekor for non-release builds? We changed it to prevent it once. Not sure if this is still the case.

Yes, every step that has `COSIGN_EXPERIMENTAL=1` will publish to Rekor. For container SBOM signing this is currently broken as the code comment details.

What is the concern about not publishing to Rekor?
I agree with Malte that using two sets of keys would be a great way to test the feature every step of the way! Should we move that to a dedicated branch? We could hold of here until the other one is merged.

**Comment by daniel-weisse** [comment]
>Yes, every step that has COSIGN_EXPERIMENTAL=1 will publish to Rekor.

Does this mean we publish measurements to Rekor even for CI runs on PRs?

**Comment by thomasten** [comment]
> What is the concern about not publishing to Rekor?

It's about https://docs.edgeless.systems/constellation/workflows/verify-cli#optional-manually-inspect-the-transparency-log, especially the last setence of this section. It makes more sense if only release artifacts are included.

> I agree with Malte that using two sets of keys would be a great way to test the feature every step of the way! Should we move that to a dedicated branch? We could hold of here until the other one is merged.

I also agree. Sounds good.

File: docs/sidebars.js
Comment (thomasten) [inline docs/sidebars.js:173]:
```suggestion
          label: 'Consume SBOMs',
```
This is quite long and produces an ugly line break in the sidebar. Do you think it's okay to shorten it to the acronym?

**Comment by datosh** [comment]
Holding off on merging this until separate key pair for non-release & release artifacts is implemented.

File: docs/docs/workflows/sbom.md
Comment (thomasten) [inline docs/docs/workflows/sbom.md:1]:
```suggestion
# Consume software bill of materials (SBOMs)
```
I find both sentence-style and title-style capitalization equally often. As MS guide prefers sentence-style, I'd suggest this.

I'm also not sure whether the plural actually needs to be *bills*?
```suggestion
# Consume software bills of materials (SBOMs)
```

File: docs/docs/workflows/sbom.md
Comment (thomasten) [inline docs/docs/workflows/sbom.md:4]:
```suggestion
Constellation builds produce a [software bill of materials (SBOM)](https://www.ntia.gov/SBOM) for each generated [artifact](../architecture/components.md).
```

File: docs/docs/workflows/sbom.md
Comment (thomasten) [inline docs/docs/workflows/sbom.md:5]:
```suggestion
You can use SBOMs to make informed decisions about dependencies and vulnerabilities in a given application. Enterprises rely on SBOMs to maintain an inventory of used applications, which allows them to take data-driven approaches to managing risks related to vulnerabilities.
```

File: docs/docs/workflows/sbom.md
Comment (thomasten) [inline docs/docs/workflows/sbom.md:7]:
```suggestion
SBOMs for Constellation are generated using [Syft](https://github.com/anchore/syft), signed using [Cosign](https://github.com/sigstore/cosign), and stored with the produced artifact.
```

**Review by thomasten — COMMENTED** [review]
some style suggestions

File: docs/sidebars.js
Comment (datosh) [reply docs/sidebars.js]: Thanks for noting! I was also unsure during writing, the shorter version is probably better for a consistently styled sidebar.

File: docs/docs/workflows/sbom.md
Comment (datosh) [reply docs/docs/workflows/sbom.md]: I wen't with the first suggestion, double plural reads odd to me.

**Review by malt3 — APPROVED** [review]
LGTM

**Pull request merged**
